File: web/src/pages/mail.ts
/**
 * Notification mail through the company SMTP relay (MAIL-STANDARD), checked
 * here the way the server checks it so a save that would be refused is said
 * before it is sent.
 */

export type MailValues = Record<string, string | number | boolean | string[]>

export const securityChoices = [
  { id: 'auto', label: '자동 — 릴레이가 STARTTLS 를 알리면 쓰고, 아니면 평문' },
  { id: 'none', label: '없음 — 평문 (사내 릴레이 25번 포트의 흔한 경우)' },
  { id: 'starttls', label: 'STARTTLS 필수' },
  { id: 'tls', label: 'TLS — 접속부터 암호화 (대개 465번 포트)' },
]

/** Says what is wrong with the section as typed, or nothing. */
export function mailProblem(values: MailValues): string {
  const host = String(values.smtp_host ?? '').trim()
  if (values.enabled === true && !host) return '메일 알림을 켜려면 SMTP 릴레이 호스트가 필요합니다.'
  if (host && (/[\s/:@]/.test(host) || host.length > 253)) return 'SMTP 호스트는 포트 없이 호스트 이름이나 주소만 적습니다.'
  const port = Number(values.smtp_port)
  if (!Number.isInteger(port) || port < 1 || port > 65535) return 'SMTP 포트는 1~65535 의 정수여야 합니다.'
  if (!securityChoices.some((choice) => choice.id === String(values.security))) return '보안 방식을 선택해 주세요.'
  const from = String(values.from_address ?? '').trim()
  if (from && (!from.includes('@') || /[\s<>,]/.test(from))) return '보내는 주소는 메일 주소 하나여야 합니다.'
  const base = String(values.base_url ?? '').trim()
  if (base && !/^https?:\/\/[^\s]+$/.test(base)) return '링크 주소는 http(s):// 로 시작하는 주소여야 합니다.'
  const seconds = Number(values.timeout_seconds)
  if (!Number.isInteger(seconds) || seconds < 1 || seconds > 120) return '제한 시간은 1~120초여야 합니다.'
  return ''
}

/** What the saved settings amount to, in one line for the top of the section. */
export function mailSummary(values: MailValues, passwordConfigured: boolean): string {
  if (values.enabled !== true) return '메일 알림이 꺼져 있습니다. 아무것도 보내지 않습니다.'
  const host = String(values.smtp_host ?? '').trim() || '(호스트 없음)'
  const port = Number(values.smtp_port) || 25
  const security = securityChoices.find((choice) => choice.id === String(values.security))?.id ?? 'auto'
  const auth = String(values.username ?? '').trim() ? `인증 ${passwordConfigured ? '있음' : '(비밀번호 없음)'}` : '인증 없음'
  return `${host}:${port} · ${security} · ${auth} 으로 보냅니다.`
}

File: server/internal/mail/mail.go
// Package mail sends event notifications through a company SMTP relay.
//
// An internal relay commonly takes mail on port 25 with no credentials and no
// TLS, so that is what the defaults describe; authentication and encryption
// are used when the relay offers them. Nothing here holds a request up: a
// notification is sent in the background, and every attempt — sent or not —
// is recorded so an administrator can answer "it never came".
package mail

import (
	"context"
	"crypto/tls"
	"errors"
	"fmt"
	"net"
	"net/smtp"
	"strings"
	"time"
)

var (
	ErrDisabled = errors.New("mail is disabled")
	ErrInvalid  = errors.New("invalid mail configuration")
)

// Config is the relay as the administrator described it.
type Config struct {
	Enabled     bool
	Host        string
	Port        int
	Security    string
	SkipVerify  bool
	Username    string
	Password    string
	FromAddress string
	FromName    string
	BaseURL     string
	Timeout     time.Duration
	// Events is the administrator's switch for each kind of notification. A
	// kind not in it is sent: adding a notification never needs a setting first.
	Events map[string]bool
}

// Allows reports whether one kind of notification is to be sent.
func (c Config) Allows(event string) bool {
	if enabled, known := c.Events[event]; known {
		return enabled
	}
	return true
}

// Address is the From header value.
func (c Config) Address() string {
	from := strings.TrimSpace(c.FromAddress)
	if name := strings.TrimSpace(c.FromName); name != "" {
		return fmt.Sprintf("%s <%s>", name, from)
	}
	return from
}

func (c Config) endpoint() string { return net.JoinHostPort(c.Host, fmt.Sprint(c.Port)) }

// Validate says why this configuration cannot send, or nil. It is what the
// settings screen refuses a save on when mail is switched on with nothing to
// send through, and what a delivery is recorded as failing on otherwise.
func (c Config) Validate() error {
	if !c.Enabled {
		return nil
	}
	return c.validate()
}

func (c Config) validate() error {
	if strings.TrimSpace(c.Host) == "" {
		return fmt.Errorf("%w: mail.smtp_host is required", ErrInvalid)
	}
	if c.Port < 1 || c.Port > 65535 {
		return fmt.Errorf("%w: mail.smtp_port must be between 1 and 65535", ErrInvalid)
	}
	if !strings.Contains(c.FromAddress, "@") {
		return fmt.Errorf("%w: mail.from_address must be an email address", ErrInvalid)
	}
	switch c.Security {
	case SecurityAuto, SecurityNone, SecurityStartTLS, SecurityTLS:
	default:
		return fmt.Errorf("%w: mail.security must be auto, none, starttls or tls", ErrInvalid)
	}
	return nil
}

// Message is one mail to one person.
type Message struct {
	To      string
	Subject string
	Body    string
}

// Sender delivers one message; the service takes a real relay by default and a
// test hands in one that remembers what it was given.
type Sender func(context.Context, Config, Message) error

// Deliver opens a connection to the relay and sends one message.
func Deliver(ctx context.Context, config Config, message Message) error {
	if err := config.validate(); err != nil {
		return err
	}
	if strings.TrimSpace(message.To) == "" {
		return fmt.Errorf("%w: recipient is required", ErrInvalid)
	}
	client, err := dial(ctx, config)
	if err != nil {
		return err
	}
	defer func() { _ = client.Close() }()
	if err := startSession(client, config); err != nil {
		return err
	}
	if err := client.Mail(strings.TrimSpace(config.FromAddress)); err != nil {
		return fmt.Errorf("MAIL FROM refused: %w", err)
	}
	if err := client.Rcpt(strings.TrimSpace(message.To)); err != nil {
		return fmt.Errorf("RCPT TO refused: %w", err)
	}
	writer, err := client.Data()
	if err != nil {
		return fmt.Errorf("DATA refused: %w", err)
	}
	if _, err := writer.Write([]byte(compose(config, message))); err != nil {
		return fmt.Errorf("body not sent: %w", err)
	}
	if err := writer.Close(); err != nil {
		return fmt.Errorf("body not accepted: %w", err)
	}
	return client.Quit()
}

func dial(ctx context.Context, config Config) (*smtp.Client, error) {
	timeout := config.Timeout
	if timeout <= 0 {
		timeout = defaultTimeout
	}
	dialer := &net.Dialer{Timeout: timeout}
	var connection net.Conn
	var err error
	if config.Security == SecurityTLS {
		connection, err = tls.DialWithDialer(dialer, "tcp", config.endpoint(), config.tlsConfig())
	} else {
		connection, err = dialer.DialContext(ctx, "tcp", config.endpoint())
	}
	if err != nil {
		return nil, fmt.Errorf("SMTP connection failed: %w", err)
	}
	// A relay that answers the connection and then says nothing would hold
	// the sender for as long as the operating system allows; every step of the
	// dialogue is bounded instead.
	_ = connection.SetDeadline(time.Now().Add(timeout))
	client, err := smtp.NewClient(connection, config.Host)
	if err != nil {
		_ = connection.Close()
		return nil, fmt.Errorf("SMTP session failed: %w", err)
	}
	return client, nil
}

// startSession upgrades and authenticates only as far as the relay offers, so
// an internal relay that takes anything and a provider that wants both work
// off the same settings.
func startSession(client *smtp.Client, config Config) error {
	if err := client.Hello(helloName(config)); err != nil {
		return fmt.Errorf("EHLO refused: %w", err)
	}
	if config.Security == SecurityStartTLS || config.Security == SecurityAuto {
		if supported, _ := client.Extension("STARTTLS"); supported {
			if err := client.StartTLS(config.tlsConfig()); err != nil {
				return fmt.Errorf("STARTTLS failed: %w", err)
			}
		} else if config.Security == SecurityStartTLS {
			return fmt.Errorf("%w: the relay does not offer STARTTLS", ErrInvalid)
		}
	}
	if strings.TrimSpace(config.Username) == "" {
		return nil
	}
	supported, mechanisms := client.Extension("AUTH")
	if !supported {
		return fmt.Errorf("%w: the relay does not offer authentication; leave the username empty", ErrInvalid)
	}
	offered := strings.ToUpper(mechanisms)
	switch {
	case strings.Contains(offered, "PLAIN"):
		return client.Auth(smtp.PlainAuth("", config.Username, config.Password, config.Host))
	case strings.Contains(offered, "LOGIN"):
		return client.Auth(loginAuth{username: config.Username, password: config.Password, host: config.Host})
	default:
		return client.Auth(smtp.CRAMMD5Auth(config.Username, config.Password))
	}
}

func (c Config) tlsConfig() *tls.Config {
	return &tls.Config{ServerName: c.Host, MinVersion: tls.VersionTLS12, InsecureSkipVerify: c.SkipVerify} //nolint:gosec // opt-in, for an internal relay with a private certificate
}

// helloName is the sender's domain, which a relay that checks the greeting
// takes better than a container hostname.
func helloName(config Config) string {
	if at := strings.LastIndex(config.FromAddress, "@"); at >= 0 && at+1 < len(config.FromAddress) {
		return config.FromAddress[at+1:]
	}
	return "localhost"
}

// loginAuth is the LOGIN mechanism, which several corporate relays offer
// instead of PLAIN; the standard library ships only PLAIN and CRAM-MD5.
type loginAuth struct{ username, password, host string }

func (a loginAuth) Start(server *smtp.ServerInfo) (string, []byte, error) {
	if !server.TLS && server.Name != a.host {
		return "", nil, errors.New("LOGIN authentication is only sent to the relay that was asked for")
	}
	return "LOGIN", nil, nil
}

func (a loginAuth) Next(fromServer []byte, more bool) ([]byte, error) {
	if !more {
		return nil, nil
	}
	switch strings.ToLower(strings.TrimRight(string(fromServer), ": ")) {
	case "username":
		return []byte(a.username), nil
	case "password":
		return []byte(a.password), nil
	}
	return nil, fmt.Errorf("unexpected LOGIN prompt %q", fromServer)
}
